File: youkai-front/src/domain/youkai/checkdigit/converter/CodeToCharacterConverter.ts
/**
 * 文字コード変換テーブル。
 */
export default class CodeToCharacterConverter {
    private readonly codeToChar: string[];
    private readonly incrementNextCodeTable: number[];

    private static readonly INVALID_CHAR = '*';
    private static readonly DEFAULT_CHAR_TABLE = "AHOV16**BIPW27**CJQX38**DKRY49**ELSZ50**FMT-n!**GNU.mc";

    constructor(
        private readonly charTable: string = CodeToCharacterConverter.DEFAULT_CHAR_TABLE,
    ) {
        this.codeToChar = charTable.split('');
        this.incrementNextCodeTable = this.createIncrementNextCodeTable(this.codeToChar);
    }

    public convert(code: number): string {
        return this.codeToChar[code];
    }

    public reverceConvert(passwordChar: string): number {
        const oneChar = passwordChar.charAt(0); // 念の為
        for (let i = 0; i < this.codeToChar.length; i++) {
            if (this.codeToChar[i] === oneChar) return i;
        }
        return -1;
    }

    public incrementCode(code: number): number {
        let index = this.incrementNextCodeTable.length - 1;
        if (code <= index) index = code;
        return this.incrementNextCodeTable[index];
    }

    public minCode(): number {
        const table = this.incrementNextCodeTable;
        return table[table.length - 1];
    }

    public isInvalidCharCode(code: number): boolean {
        return CodeToCharacterConverter.INVALID_CHAR === this.convert(code);
    }

    public isInvalidChar(oneChar: string): boolean {
        if (CodeToCharacterConverter.equalInvalidChar(oneChar)) return true;
        return !this.charTable.includes(oneChar);
    }

    public static equalInvalidChar(oneChar: string): boolean {
        return oneChar === CodeToCharacterConverter.INVALID_CHAR;
    }

    public isInvalidPassword(text: string): boolean {
        if (text.trim().length === 0) return false;
        return text.split("")
            .some(oneChar => this.isInvalidChar(oneChar));
    }

    public fixValidPassword(text: string): string {
        if (text.trim().length === 0) return "";
        return text.split("")
            .map(oneChar => this.fixValidOneChar(oneChar))
            .join("");
    }

    private fixValidOneChar(oneChar: string) {
        if (!this.isInvalidChar(oneChar)) return oneChar;
        const upperChar = oneChar.toUpperCase();
        if (!this.isInvalidChar(upperChar)) return upperChar;
        return "";
    }

    public validCharacters(): string {
        return this.charTable
            .split("")
            .filter(c => c !== CodeToCharacterConverter.INVALID_CHAR)
            .sort()
            .join("");
    }

    private createIncrementNextCodeTable(codeToChar: string[]): number[] {
        const firstValidCharCode = [...Array(codeToChar.length).keys()]
            .filter(i => !CodeToCharacterConverter.equalInvalidChar(codeToChar[i]))[0];

        const result = Array(codeToChar.length);
        let currentIncrementCode = firstValidCharCode;
        for (let i = codeToChar.length - 1; i >= 0; i--) {
            result[i] = currentIncrementCode;
            if (!CodeToCharacterConverter.equalInvalidChar(codeToChar[i])) currentIncrementCode = i;
        }

        return result;
    }
}

File: youkai-core/src/main/java/com/github/kazuhitom/youkai/core/domain/model/password/converter/CodeToCharacterConverter.java
package com.github.kazuhitom.youkai.core.domain.model.password.converter;

import java.util.Arrays;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static java.util.stream.Collectors.joining;

/**
 * 文字コード変換テーブル。
 */
public class CodeToCharacterConverter {
    private final String charTable;
    private final char[] codeToChar;
    private final int[] incrementNextCodeTable;

    private static final char INVALID_CHAR = '*';
    private static final String DEFAULT_CHAR_TABLE = "AHOV16**BIPW27**CJQX38**DKRY49**ELSZ50**FMT-n!**GNU.mc";

    public char convert(int code) {
        return codeToChar[code];
    }

    public int reverceConvert(char oneChar) {
        for (int i = 0; i < this.codeToChar.length; i++) {
            if (codeToChar[i] == oneChar) return i;
        }
        return -1;
    }

    public int reverceConvert(String passwordChar) {
        int oneChar = passwordChar.charAt(0); // 念の為
        for (int i = 0; i < this.codeToChar.length; i++) {
            if (this.codeToChar[i] == oneChar) return i;
        }
        return -1;
    }


    public int incrementCode(int code) {
        int index = incrementNextCodeTable.length - 1;
        if (code <= index) index = code;
        return incrementNextCodeTable[index];
    }

    public int minCode() {
        return incrementNextCodeTable[incrementNextCodeTable.length - 1];
    }

    public int maxCode() {
        return Arrays.stream(this.incrementNextCodeTable)
                .min()
                .getAsInt();
    }

    public boolean isInvalidCharCode(int code) {
        return INVALID_CHAR == convert(code);
    }


    public boolean isInvalidChar(String oneChar) {
        if (CodeToCharacterConverter.equalInvalidChar(oneChar)) return true;
        return !this.charTable.contains(oneChar);
    }

    private int[] createIncrementNextCodeTable(char[] codeToChar) {
        int firstValidCharCode = IntStream.range(0, codeToChar.length)
                .filter(i -> codeToChar[i] != INVALID_CHAR)
                .findFirst()
                .orElse(0);

        int[] result = new int[codeToChar.length];
        int currentIncrementCode = firstValidCharCode;
        for (int i = codeToChar.length - 1; i >= 0; i--) {
            result[i] = currentIncrementCode;
            if (codeToChar[i] != INVALID_CHAR) currentIncrementCode = i;
        }

        return result;
    }

    public int randomCode() {
        String validChars = this.validCharacters();
        int randomPos = (int) Math.floor(Math.random() * validChars.length());
        String oneChar = validChars.substring(randomPos, 1);
        return this.reverceConvert(oneChar);
    }

    public static boolean equalInvalidChar(String oneChar) {
        return oneChar == String.valueOf(CodeToCharacterConverter.INVALID_CHAR);
    }

    public boolean isInvalidPassword(String text) {
        if (text.trim().length() == 0) return false;
        return Arrays.stream(text.split(""))
                .anyMatch(oneChar -> isInvalidChar(oneChar));
    }

    public String fixValidPassword(String text) {
        if (text.trim().length() == 0) return "";
        return Arrays.stream(text.split(""))
                .map(oneChar -> fixValidOneChar(oneChar))
                .collect(joining());
    }

    private String fixValidOneChar(String oneChar) {
        if (!this.isInvalidChar(oneChar)) return oneChar;
        String upperChar = oneChar.toUpperCase();
        if (!this.isInvalidChar(upperChar)) return upperChar;
        return "";
    }

    public String validCharacters() {
        String invalidCharString = String.valueOf(INVALID_CHAR);
        return Stream.of(charTable.split(""))
                .filter(c -> !c.equals(invalidCharString))
                .collect(joining());
    }

    public CodeToCharacterConverter() {
        this(DEFAULT_CHAR_TABLE);
    }

    public CodeToCharacterConverter(String charTable) {
        this.charTable = charTable;
        this.codeToChar = charTable.toCharArray();
        this.incrementNextCodeTable = createIncrementNextCodeTable(this.codeToChar);
    }
}
